Give DirectoryPackage a strong-name

jbaehr opened the pull request:
Having an assembly strong-named is a pre-condition to reference it from other strong-named assemblies.
The signing key was generated using Visual Studio 2019 an is not password protected, as per recommendation for open source projects.

See also
- https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named

Note that strong-naming an assembly may require recompilation of a consuming app that used to consume the simple-named version. So you may consider this a breaking change. However, as far as I see DirectoryPackage never had an official release, so this may not be a big concern.

jbaehr committed:
Give DirectoryPackage a strong-name

Having an assembly strong-named is a pre-condition to reference it from
other strong-named assemblies.
The signing key is not password protected, as per recommendation for
open source projects.
See also
- https://docs.microsoft.com/en-us/dotnet/standard/assembly/strong-named

andersforsgren commented:
Hi,

I'm happy to add StrongNaming and there is no need to worry about existing users/breaking changes. But...

I added GitVersion automatic versioning now, (hence the conflict, sorry...)

The GitVersion docs says this about StrongNaming

_**assembly-versioning-scheme** 
When updating assembly info, assembly-versioning-scheme tells GitVersion how to treat the AssemblyVersion attribute. Useful to lock the major when using Strong Naming. Note: you can use None to skip updating the AssemblyVersion while still updating the AssemblyFileVersion and AssemblyInformationVersion attributes. Valid values: MajorMinorPatchTag, MajorMinorPatch, MajorMinor, Major, None._

https://gitversion.net/docs/reference/configuration

I don't  fully understand that, i.e. which configuration of GitVersion should be used when Strong Naming? If you do grok this, please update the GitVersion yml appropriately when you rebase this PR, thanks.

jbaehr committed:
Adaprt GitVersion.yml to play nice with strong-naming and SemVer

To be able to replace a dependency without recompilation of the
consumer, the `AssemblyVersion` must not change (otherwise, binding
redicects are required). Following the rules of Semantic Versioning,
there are no breaking changes within the same major. Thus we only
change/increment the Assembly Version in case of major-bumb.

For Windows Installer to replace a file on disk during an upgrade, the
`AssemblyFileVersion` has to be larger. Thus we have to ensure newer
versions have a larger value here. We cannot naively use the pre-release
number here as this one is reseted then changing the pre-release label.
E.g. after "alpha.3" comes "beta.1" -- non the less the beta has to
replace the alpha. For this to happen, we put a "weight" on the
prerelease, as the file version only allows integers. The weight is
choosen to still for in 16 bit *and* make it easy for humans to tell
them apart.

Last but not least, the mainline increment is changed from "patch" to
"minor" as I expect that the master branch heads for new feature
releases by default. For occasional hotfixes, only incrementing patch, a
dedicated branch may be used.

jbaehr commented:
> I don't fully understand that, i.e. which configuration of GitVersion should be used when Strong Naming? If you do grok this, please update the GitVersion yml appropriately when you rebase this PR, thanks.

Here you go. I replaced the last commit with a new one adapting the GitVersion.yml
